<a href="https://colab.research.google.com/github/MoseAIML/Moses-/blob/main/MarketBasketAnalysis.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Background and Problem Statement Care five is a German multinational retail corporation headquartered in Berlin, Germany. It is the eighth-largest retailer in the world by revenue. It operates a chain of hypermarkets, groceries stores, and convenience stores, which as of January 2021, comprises its 12,00 stores in over 30 countries. As a Data analyst working for one of the stores, you must perform market basket analysis to help the store maximize revenue. More specifically, your task will analyze transactional data to identify the top 10 products likely to be purchased together.

In [1]:
# Import the required libraries
import pandas as pd
import numpy as np               
import matplotlib.pyplot as plt
from mlxtend.frequent_patterns import apriori
from mlxtend.frequent_patterns import association_rules

In [2]:
df = pd.read_csv("https://bit.ly/30A2gHO")
df.head(10)

A  Quantity  Transaction  Store    Product
0  30000         2        93194      6   Magazine
1  30001         2        93194      6  Candy Bar
2  30002         2        93194      6  Candy Bar
3  30003         2        93194      6  Candy Bar
4  30004         2        93194      6  Candy Bar
5  30005         2        93197      1    Pencils
6  30006         1        93200      6  Candy Bar
7  30007         1        93200      6  Candy Bar
8  30008         1        93200      6  Candy Bar
9  30009         1        93200      6   Magazine

In [3]:
#Group the basket dataframe by Transaction & Product
#Display the count of items
# ---
df2 = df.groupby(['Transaction','Product']).size().reset_index(name='Count')
df2.head(10)

Transaction         Product  Count
0        93194       Candy Bar      4
1        93194        Magazine      1
2        93197         Pencils      1
3        93200       Candy Bar      3
4        93200        Magazine      1
5        93206  Greeting Cards      1
6        93206        Magazine      1
7        93206         Pencils      2
8        93212      Toothbrush      1
9        93215       Candy Bar      2

In [4]:
#Consolidate the items into one transaction per row 
#One Hot Encode
# ---
#
df3 = (df2.groupby(['Transaction', 'Product'])['Count']
          .sum().unstack().reset_index().fillna(0)
          .set_index('Transaction'))

df3.head(10)

Product      Bow  Candy Bar  Deodorant  Greeting Cards  Magazine  Markers  \
Transaction                                                                 
93194        0.0        4.0        0.0             0.0       1.0      0.0   
93197        0.0        0.0        0.0             0.0       0.0      0.0   
93200        0.0        3.0        0.0             0.0       1.0      0.0   
93206        0.0        0.0        0.0             1.0       1.0      0.0   
93212        0.0        0.0        0.0             0.0       0.0      0.0   
93215        0.0        2.0        0.0             1.0       0.0      0.0   
93233        0.0        0.0        0.0             0.0       0.0      0.0   
93239        0.0        0.0        0.0             0.0       1.0      0.0   
93245        0.0        1.0        0.0             0.0       0.0      0.0   
93248        0.0        1.0        0.0             0.0       0.0      0.0   

Product      Pain Reliever  Pencils  Pens  Perfume  Photo Processing  \
Transaction                                                            
93194                  0.0      0.0   0.0      0.0               0.0   
93197                  0.0      1.0   0.0      0.0               0.0   
93200                  0.0      0.0   0.0      0.0               0.0   
93206                  0.0      2.0   0.0      0.0               0.0   
93212                  0.0      0.0   0.0      0.0               0.0   
93215                  0.0      0.0   0.0      0.0               0.0   
93233                  0.0      0.0   0.0      0.0               0.0   
93239                  0.0      0.0   0.0      0.0               0.0   
93245                  0.0      0.0   0.0      0.0               0.0   
93248                  0.0      0.0   0.0      0.0               0.0   

Product      Prescription Med  Shampoo  Soap  Toothbrush  Toothpaste  \
Transaction                                                            
93194                     0.0      0.0   0.0         0.0         0.0   
93197                     0.0      0.0   0.0         0.0         0.0   
93200                     0.0      0.0   0.0         0.0         0.0   
93206                     0.0      0.0   0.0         0.0         0.0   
93212                     0.0      0.0   0.0         1.0         0.0   
93215                     0.0      0.0   0.0         0.0         0.0   
93233                     0.0      0.0   0.0         1.0         0.0   
93239                     0.0      0.0   0.0         0.0         0.0   
93245                     0.0      0.0   0.0         0.0         0.0   
93248                     0.0      0.0   0.0         0.0         0.0   

Product      Wrapping Paper  
Transaction                  
93194                   0.0  
93197                   0.0  
93200                   0.0  
93206                   0.0  
93212                   0.0  
93215                   0.0  
93233                   0.0  
93239                   0.0  
93245                   0.0  
93248                   0.0

In [5]:
#Use encoding function to convert all the values to 0 or 1. 
#Apriori algorithm will only take 0's or 1's.
# ---
# 
def encode_units(x):
    if x <= 0:
        return 0
    if x >= 1:
        return 1

df4 = df3.applymap(encode_units)

df4.head()

Product      Bow  Candy Bar  Deodorant  Greeting Cards  Magazine  Markers  \
Transaction                                                                 
93194          0          1          0               0         1        0   
93197          0          0          0               0         0        0   
93200          0          1          0               0         1        0   
93206          0          0          0               1         1        0   
93212          0          0          0               0         0        0   

Product      Pain Reliever  Pencils  Pens  Perfume  Photo Processing  \
Transaction                                                            
93194                    0        0     0        0                 0   
93197                    0        1     0        0                 0   
93200                    0        0     0        0                 0   
93206                    0        1     0        0                 0   
93212                    0        0     0        0                 0   

Product      Prescription Med  Shampoo  Soap  Toothbrush  Toothpaste  \
Transaction                                                            
93194                       0        0     0           0           0   
93197                       0        0     0           0           0   
93200                       0        0     0           0           0   
93206                       0        0     0           0           0   
93212                       0        0     0           1           0   

Product      Wrapping Paper  
Transaction                  
93194                     0  
93197                     0  
93200                     0  
93206                     0  
93212                     0

In [6]:
# Frequent Items
shop_frequent_itemsets = apriori(df4, min_support=0.01, use_colnames=True)
shop_frequent_itemsets.head(15)

support            itemsets
0   0.051591               (Bow)
1   0.175736         (Candy Bar)
2   0.152840    (Greeting Cards)
3   0.231936          (Magazine)
4   0.020071     (Pain Reliever)
5   0.135147           (Pencils)
6   0.144068              (Pens)
7   0.082664           (Perfume)
8   0.055456  (Photo Processing)
9   0.014422  (Prescription Med)
10  0.029289           (Shampoo)
11  0.039102              (Soap)
12  0.067648        (Toothbrush)
13  0.162504        (Toothpaste)
14  0.060065    (Wrapping Paper)

In [7]:
# Generate association rules
#Finding the association rules
shop_rules = association_rules(shop_frequent_itemsets, metric="lift", min_threshold=1)

# Sorting 
shop_rules.sort_values("lift", ascending = False, inplace = True)

#Top 10 combinations of items frequently bought together
shop_rules.head(15)

antecedents                   consequents  \
19                   (Perfume)                  (Toothbrush)   
18                (Toothbrush)                     (Perfume)   
0                        (Bow)                  (Toothbrush)   
1                 (Toothbrush)                         (Bow)   
24            (Greeting Cards)         (Candy Bar, Magazine)   
21       (Candy Bar, Magazine)              (Greeting Cards)   
54            (Greeting Cards)           (Pencils, Magazine)   
51         (Pencils, Magazine)              (Greeting Cards)   
47                 (Candy Bar)         (Pencils, Toothpaste)   
46       (Pencils, Toothpaste)                   (Candy Bar)   
23                 (Candy Bar)    (Greeting Cards, Magazine)   
22  (Greeting Cards, Magazine)                   (Candy Bar)   
41                 (Candy Bar)        (Magazine, Toothpaste)   
40      (Magazine, Toothpaste)                   (Candy Bar)   
29                 (Candy Bar)  (Greeting Cards, Toothpaste)   

    antecedent support  consequent support   support  confidence      lift  \
19            0.082664            0.067648  0.017098    0.206835  3.057514   
18            0.067648            0.082664  0.017098    0.252747  3.057514   
0             0.051591            0.067648  0.010110    0.195965  2.896843   
1             0.067648            0.051591  0.010110    0.149451  2.896843   
24            0.152840            0.039994  0.017247    0.112840  2.821431   
21            0.039994            0.152840  0.017247    0.431227  2.821431   
54            0.152840            0.028546  0.012043    0.078794  2.760244   
51            0.028546            0.152840  0.012043    0.421875  2.760244   
47            0.175736            0.022748  0.011002    0.062606  2.752198   
46            0.022748            0.175736  0.011002    0.483660  2.752198   
23            0.175736            0.037467  0.017247    0.098139  2.619370   
22            0.037467            0.175736  0.017247    0.460317  2.619370   
41            0.175736            0.029884  0.013232    0.075296  2.519610   
40            0.029884            0.175736  0.013232    0.442786  2.519610   
29            0.175736            0.033304  0.014570    0.082910  2.489530   

    leverage  conviction  
19  0.011506    1.175482  
18  0.011506    1.227611  
0   0.006620    1.159592  
1   0.006620    1.115055  
24  0.011134    1.082112  
21  0.011134    1.489452  
54  0.007680    1.054546  
51  0.007680    1.465358  
47  0.007005    1.042520  
46  0.007005    1.596359  
23  0.010662    1.067274  
22  0.010662    1.527313  
41  0.007981    1.049110  
40  0.007981    1.479260  
29  0.008718    1.054091

In [8]:
# Metric interpretation
# Sorting 
shop_rules.sort_values("confidence", ascending = False, inplace = True)

# Previewing the associative rules
shop_rules.head(15)

antecedents       consequents  antecedent support  \
46         (Pencils, Toothpaste)       (Candy Bar)            0.022748   
22    (Greeting Cards, Magazine)       (Candy Bar)            0.037467   
40        (Magazine, Toothpaste)       (Candy Bar)            0.029884   
28  (Greeting Cards, Toothpaste)       (Candy Bar)            0.033304   
21         (Candy Bar, Magazine)  (Greeting Cards)            0.039994   
51           (Pencils, Magazine)  (Greeting Cards)            0.028546   
50     (Greeting Cards, Pencils)        (Magazine)            0.029884   
20   (Candy Bar, Greeting Cards)        (Magazine)            0.046090   
58        (Magazine, Toothpaste)  (Greeting Cards)            0.029884   
34           (Magazine, Pencils)       (Candy Bar)            0.028546   
27       (Candy Bar, Toothpaste)  (Greeting Cards)            0.041481   
56  (Greeting Cards, Toothpaste)        (Magazine)            0.033304   
38         (Candy Bar, Magazine)      (Toothpaste)            0.039994   
52    (Greeting Cards, Magazine)         (Pencils)            0.037467   
39       (Candy Bar, Toothpaste)        (Magazine)            0.041481   

    consequent support   support  confidence      lift  leverage  conviction  
46            0.175736  0.011002    0.483660  2.752198  0.007005    1.596359  
22            0.175736  0.017247    0.460317  2.619370  0.010662    1.527313  
40            0.175736  0.013232    0.442786  2.519610  0.007981    1.479260  
28            0.175736  0.014570    0.437500  2.489530  0.008718    1.465358  
21            0.152840  0.017247    0.431227  2.821431  0.011134    1.489452  
51            0.152840  0.012043    0.421875  2.760244  0.007680    1.465358  
50            0.231936  0.012043    0.402985  1.737486  0.005112    1.286508  
20            0.231936  0.017247    0.374194  1.613350  0.006557    1.227319  
58            0.152840  0.011151    0.373134  2.441344  0.006583    1.351422  
34            0.175736  0.010407    0.364583  2.074609  0.005391    1.297202  
27            0.152840  0.014570    0.351254  2.298188  0.008230    1.305844  
56            0.231936  0.011151    0.334821  1.443595  0.003426    1.154674  
38            0.162504  0.013232    0.330855  2.035984  0.006733    1.251592  
52            0.135147  0.012043    0.321429  2.378359  0.006979    1.274520  
39            0.231936  0.013232    0.318996  1.375365  0.003611    1.127842

Observation & Recommendation Highest Lift

Toothbrush and Perfume are the items most likely to be purchased together with the highest lift of 3.057514. They should be palaced close to each. Highest confidence levels

The purchase of Toothpaste, Pencils, Magazine, Greeting Cards and Candy Bar go hand in hand. The Supermarket should consider placing the above items on the same aisle.